### ISSUE TITLE
Cyclical features when variables are negative - does it make sense?

### ISSUE THREAD

**Issue by solegalli** [feature-engine/feature_engine#578]
Not sure we can make a variable with negative values cyclical with the sine and cosine. We need to investigate more and if it does not make sense, we need to put a safeguard.

**Comment by VascoSch92** [comment]
I think the main problem is: what happen if the max value is 0? An error is thrown? 

In this case what we could also do is taking the max of the absolute value of the column and still having an encoding which make sense.

**Comment by VascoSch92** [comment]
Yes yes you are right. I was just speaking about the case where the max of the column was 0.

From what i understand: you would like to find a solution that doesn't differentiate cases. Right?

I'm thinking that you could always translate the column by the max of the absolute value of the columns. This should not change the cyclical embedding as sin and cos are periodic functions.

But i have to check this last sentence ;-)

**Comment by VascoSch92** [comment]
I think about the problem and this is my conclusion:

There is a good and intelligent way to get rid of the problem of negative values. It suffices to take 
`max_value = | min(column) - max(column) |`

Because given a cyclical features, we are searching the length of the domain of the cyclic feature such that then we can embed it in the 2-dim periodic function (sin(...), cos(...)). The length of the domain is necessary to embed the column bijectively into the circle. 

In fact, this Is what we were already doing taking the max of the column. We were just supposing that the min was 0.

**Comment by solegalli** [comment]
Hey @VascoSch92 

I am not sure I understand the solution.

We basically want to squeze the variable between 0 and 1, so that when we multiply by 2pi it is between 0 and 2pi.

So far, the transformer assumes that the minimum is always 0 and the maximum is always positive, so we divide by the max, that squeezes the variable between 0 and 1, and then we multiply by 2Pi.

If the variable has negative and positive values, like say it ranges from -5 to 10, then we want -5 to be taken to 0 and 10 to be taken to 1, in a way that does not alter the distribution.

One option could be to add the minimum value to the entire distribution and then find the maximum and use the transformer as it is used now.

Another option could be to scale to the min and max like the MinMaxScaler:
```
(X - X.min(axis=0)) / (X.max(axis=0) - X.min(axis=0)
```

Whatever, we do, ideally, we do not want to break backwards compatibility.

Does your solution return the results I expect / describe here as well?

**Comment by VascoSch92** [comment]
Hey @solegalli 


```
We basically want to squeze the variable between 0 and 1, so that when we multiply by 2pi it is between 0 and 2pi.

So far, the transformer assumes that the minimum is always 0 and the maximum is always positive, so we divide by the max, that squeezes the variable between 0 and 1, and then we multiply by 2Pi.
``` 

Exactly, we want to do that because the interval [0, 2Pi] is the period of  `sin` and `cos`. 

```
If the variable has negative and positive values, like say it ranges from -5 to 10, then we want -5 to be taken to 0 and 10 to be taken to 1, in a way that does not alter the distribution.
```

If you look at my suggestion we have that: `| min(column) - max(column) | = - 5 - 10 = 5`. Therefore, -5 goes to 0 and 10 goes to 2. Therefore, 10 will be mapped to `(sin(4Pi), cos(4Pi)) =  (sin(2Pi), cos(2Pi))` because of periodicity (and the same for other values)

```
Whatever, we do, ideally, we do not want to break backwards compatibility.

Does your solution return the results I expect / describe here as well?
```
I understand your point and, of course, the idea is to have a user-friendly (not overcomplicated) transformer. 

Therefore, my suggestion is just to check if the columns is positive and raise an error otherwise
